src/lib/chat/matrix-client.ts

```diff
@@ -450,10 +450,15 @@ export class MatrixClient implements IChatClient {
     }
   }
 
-  private initializeRoomEventHandlers(room) {
-    room.on(RoomEvent.UnreadNotifications, (unreadNotifications) => {
+  private initializeRoomEventHandlers(room: Room) {
+    const handleUnreadNotifications = (unreadNotifications) => {
       this.events.receiveUnreadCount(room.roomId, unreadNotifications?.total || 0);
-    });
+    };
+
+    // Removing the event handler if it already exists to prevent duplicates.
+    room.off(RoomEvent.UnreadNotifications, handleUnreadNotifications);
```

Comment:
Close but this won't actually work because you're creating a brand new function every time this is called so they won't be detected as the same thing and the `off` won't do anything.

Reply:
Thanks for the insight. I'm thinking of mapping event handlers to room IDs to reuse them instead of creating new ones each time. On disconnect, we'd clean up by removing the handlers from the map. Uncertain if this would really work for cleaning them up on disconnect though.

```
  private async initializeRoomEventHandlers(room: Room) {
    console.log(this.eventHandlers);
    const roomId = room.roomId;

    let handleUnreadNotifications = this.eventHandlers.get(roomId);

    if (!handleUnreadNotifications) {
      handleUnreadNotifications = (unreadNotifications) => {
        this.events.receiveUnreadCount(roomId, unreadNotifications?.total || 0);
      };

      this.eventHandlers.set(roomId, handleUnreadNotifications);
    }
   console.log('handleUnreadNotifications', handleUnreadNotifications)
    if (!room.listenerCount(RoomEvent.UnreadNotifications)) {
      room.on(RoomEvent.UnreadNotifications, handleUnreadNotifications);
    }
  }
  ```

Reply:
In the end, I just checked the cache to see if it already existed: https://github.com/zer0-os/zOS/pull/1139/commits/249c9b81ee2f921550c3b5ff3f2ee757407f701e